### ISSUE TITLE
Consider reporting on usage of libraries that are now deprecated

### ISSUE THREAD

**Issue by lread** [clj-kondo/clj-kondo#1591]
To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

**Is your feature request related to a problem? Please describe.**
I don't know how many times I've seen @seancorfield remind folks on Slack that clj-time is now deprecated. It might be nice if clj-kondo could report cases where people use a library that is now deprecated.

**Describe the solution you'd like**
My initial thinking was to have clj-kondo warn on the usage of namespaces from deprecated libraries.

To consider:
- optionally providing advice on what to use instead
- allowing suppression of warning for all or specific deprecated libraries
- allowing user to specify their deprecated libraries config 

**Describe alternatives you've considered**
Another idea is to warn on reference to deprecated libraries from `deps.edn` (and maybe `bb.edn`, and maybe `project.clj`, and maybe this is more suited to [antq](https://github.com/liquidz/antq)?).

Even if deps warnings are better handled through a tool like antq, the clj-kondo warnings are still valuable because clj-kondo is more live (active in editor integrations, etc).

It might be interesting to consider a collaboration between clj-kondo and antq. 
Perhaps they could draw from the same source-of-truth for deprecated libraries.

**Additional context**
- Ideally, a deprecated library is one that has been clearly marked as such in its README or description.
- Once a library is deprecated, it's current and all previous versions are deprecated.
- A deprecated library is often archived and will not be getting any more updates (even to mark itself somehow as deprecated)
- Related: #1125

**Comment by seancorfield** [comment]
That's partly why I think I'd prefer this to be on `deps.edn`/`project.clj` instead of the code level @serioga so that someone could fork the library and maintain it themselves (there was discussion at one point of creating a `clj-time`-compatible wrapper for Java Time as a migration aid).

**Comment by liquidz** [comment]
@lread Thanks for your guide.
I think it's a good idea to warn about deprecated libraries in antq.

> optionally providing advice on what to use instead

The information about what to recommend instead of deprecated libraries is hard to maintain and can become noise in the future.
So in my opinion, it is better to maintain only the list of deprecated libraries in common.

**Comment by lread** [comment]
Thanks for following up @liquidz, much appreciated!

> The information about what to recommend instead of deprecated libraries is hard to maintain and can become noise in the future.

This is a very good point and is thoughtfully forward-thinking.
If we were to want to include a list of recommendations we'd need a long-term plan for who and how. @seancorfield whatcha think about that?

**Comment by lread** [comment]
Just a side thought (probably more antq than kondo): would this also handle cases where a single release has been found to be problematic but the maintainers have not gotten a chance to release a fix yet?

In my limited experience, this has happened for me for ClojureScript, kaocha, and figwheel-main.

**Comment by seancorfield** [comment]
I think a small "library" that just contains an EDN file containing library coordinates and a deprecation message would work.

It could be more sophisticated. As noted in the clj-kondo issue, some libraries may well have multiple coordinates that should all be treated the same so maybe have a canonical entry and a way to alias other coordinates to that.

Also, per @lread it might be useful to have some sort of version range that represents "deprecation" or some other problems. The default would be "all versions".

So I guess the `clj-easy` project (if that's where it ends up) would also need at least one function that, given a "lib name" and a version string returns either `nil` (no deprecation match) or a string (the deprecation/problem description).

Maintainers could then submit PRs against that repo to have problematic libraries/versions included and both `antq` and `clj-kondo` could just update to the latest version of it whenever they make a release?

**Comment by borkdude** [comment]
Indeed, that would be the idea.

Can we do it as much as possible in EDN so users can override and add to this in their clj-kondo config?
I prefer to not depend on this library at runtime, but just copy the EDN file in a build step into the clj-kondo jar/image.

**Comment by lread** [comment]
Clarification [clj-easy is an org](https://github.com/clj-easy). 
We would create a new `lib-deprecations` repo under that.

**Comment by lread** [comment]
So, I guess antq and clj-kondo release often enough to not need to fetch the latest live deprecations outside of a release, yeah?

**Comment by lread** [comment]
Can create repo, is `lib-deprecations` a good name? 
@seancorfield would love for you to be a maintainer too, interested?

**Comment by seancorfield** [comment]
Yup, sure. @borkdude The function I was proposing was just a convenience (to handle version matching and lib aliases) -- it would be very minimal code 😄

**Comment by lread** [comment]
A fresh repo is ready: https://github.com/clj-easy/lib-deprecations

**Comment by borkdude** [comment]
Added build setup based on other clj-easy projects and project is auto-deploying to clojars already (even though there's not much there yet):

https://clojars.org/com.github.clj-easy/lib-deprecations

**Comment by lread** [comment]
Another scenario I'm not sure we've explicitly covered is libraries that moved to the new clojars group name scheme.

For example: `lread/test-doc-blocks` is now `com.github.lread/test-doc-blocks`

This might be better handled at clojars and/or via antq but bring it up here for completeness.

**Comment by seancorfield** [comment]
I think that could also be handled by this machinery, since the simplest form is `lib/name` => `:message "..."` which could point users to the new repo?

**Comment by borkdude** [comment]
Agreed. Rather than lib-deprecations we could generalize to `lib/name => {:level ... :warning ...}`. There might be other interesting warnings to say about libraries: e.g. upgrade because version 0.4.0 has a serious bug in it. Hence maybe `lib-warnings` or so is a better name.

**Comment by lread** [comment]
That makes me wonder if clojars might be another tool/service that takes an interest in this initiative too. I'll be so bold as to link in @tcrawley.

**Comment by lread** [comment]
@borkdude I think your idea of renaming to `lib-warnings` makes good sense.

Some rambly brainstormy thoughts:

I suppose we should add a blurb to the README on how changes get into the database.
There might be cases where there are differences of opinion.

There won't be any contention around new clojars coordinates due to group-id changes. The artifact has moved.

But if, for example, I find an issue in a library that I feel is a show stopper, but the library author feels meh or less strongly about... I suppose we could still add a warning and the message could link to the issue? I guess I'm hinting at etiquette of working with the library author when adding a new entry. 

For abandoned libraries, I suppose there is no one to ask. 
I think I'd feel comfortable adding a recommendation to migrate to rewrite-clj v1 from rewrite-cljs. But I think I'd still ping the original author as a courtesy.

@borkdude, I think you were hinting at something with `:level`. 
I wonder what the values would be and convey.
Perhaps a `:type` would be helpful? `:relocation` `:serious-bug` `:security` `:replaced`...

And I suppose a single library could have multiple active warnings.

**Comment by borkdude** [comment]
About renaming: I was just going to post that I changed my mind already, since recommending to not use a library is in a sense deprecating it (even if it's subjective).

Adding the `:type` seems good, but maybe it's also good to start with the least amount of stuff that we need right now and not overthink it too much. Making the value of the lib/name entry a vector instead of a map seems good to be able to accommodate multiple warnings.

``` clojure
{clj-time/clj-time [{:type :deprecated :message "Use java.time instead"}]}
```

**Comment by borkdude** [comment]
I think it would be good to start with non-subjective data, i.e. confirmed by the library author. If a library is unmaintained can only be confirmed by an author and if there isn't a maintained alternative, the library is effectively not deprecated yet.

**Comment by lread** [comment]
> About renaming: I was just going to post that I changed my mind already, since recommending to not use a library is in a sense deprecating it (even if it's subjective).

Ha! Ok, fine by me too.

> Whether a library is unmaintained can only be confirmed by an author and if there isn't a maintained alternative, the library is effectively not deprecated yet.

I almost had an odd situation. I was working on a maintained version of rewrite-clj and could not reach the original author for the longest time. It all worked out in the end, but suppose I never did manage to reach them. In this type of case, maybe we would not be comfortable adding an entry.

And there are cases, sadly, where the original author is no longer with us.
An example in this realm is `Raynes/fs` which is now in maintenance mode at `clj-commons/fs` whose README suggests considering `babashka/fs` as a maintained alternative.

**Comment by borkdude** [comment]
I was also thinking about the latter case. In case of death, we have objective confirmation that the library is no longer maintained _by the author_, but even in this case, the maintainer might have appointed a follow up. Github also has a feature to transfer a repo in case of death. So even then, it's not always the case that the library will be unmaintained. We can address that case when we need to, without speculating over it now.

**Comment by borkdude** [comment]
Something I'd like to add myself in clj-kondo:

- Warn on usage of `borkdude/sci {:mvn/version "whatever"}` since SCI has moved to `org.babashka/sci` and since clj-kondo relies on `org.babashka/sci` but other projects sometimes also still use `borkdude/sci`, this has given unexpected problems. I have tried pro-actively sending PRs to move people over to the new org, but I think having that inside the linter would also be good. What I would like to see here is:

``` clojure
:deps {borkdude/sci ...}
            ^--- borkdude/sci is deprecated: library has moved org.babashka/sci
``` 

I think I'll just hard-code this for now and we can always switch to a commonly shared repo later.
